servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/TransportObserverInitializer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.transport.netty.internal;
+
+import io.servicetalk.transport.api.ConnectionObserver;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+import static io.servicetalk.transport.netty.internal.TransportObserverUtils.connectionObserver;
+
+/**
+ * A {@link ChannelInitializer} that registers a {@link ConnectionObserver} for all channels.
+ */
+public final class TransportObserverInitializer implements ChannelInitializer {
+
+    public static final ChannelInitializer TRANSPORT_OBSERVER_INITIALIZER = new TransportObserverInitializer();
+
+    private TransportObserverInitializer() {
+        // Singleton
+    }
+
+    @Override
+    public void init(final Channel channel) {
+        final ConnectionObserver observer = connectionObserver(channel);
```

Comment:
You are decoupling observer creation from channel initialization. Is the intent here to account for time delta between start of connect and call to `init()` here? 
The init sequence of creating an observer and adding the transport handler seems convoluted. Have you thought about creating the observer in this method?

Reply:
Correct, the intend was to create a `ConnectionObserver` as soon as we have access to the `Channel`. But you are right, there is not much between the netty's `initChannel` and the current `init` method. Moved it to `TransportObserverInitializer` and it looks much better. Thanks!